src/googleclouddebugger/firebase_client.py

```diff
@@ -0,0 +1,587 @@
+              f'cdbg/breakpoints/{self._debuggee_id}/snapshots/{bp_id}')
+          bp_ref.set(breakpoint_data)
+
+          # Now strip potential snapshot data.
+          breakpoint_data.pop('evaluatedExpressions', None)
+          breakpoint_data.pop('stackFrames', None)
+          breakpoint_data.pop('variableTable', None)
+
+        # Then add it to the list of final breakpoints.
+        bp_ref = firebase_admin.db.reference(
+            f'cdbg/breakpoints/{self._debuggee_id}/final/{bp_id}')
+        bp_ref.set(breakpoint_data)
+
+        native.LogInfo(f'Breakpoint {bp_id} update transmitted successfully')
+
+      # TODO: Add any firebase-related error handling.
+      except googleapiclient.errors.HttpError as err:
+        # Treat 400 error codes (except timeout) as application error that will
+        # not be retried. All other errors are assumed to be transient.
+        status = err.resp.status
+        is_transient = ((status >= 500) or (status == 408))
+        if is_transient:
+          if retry_count < self.max_transmit_attempts - 1:
+            native.LogInfo(f'Failed to send breakpoint {bp_id} update: '
+                           f'{traceback.format_exc()}')
+            retry_list.append((breakpoint_data, retry_count + 1))
+          else:
+            native.LogWarning(
+                f'Breakpoint {bp_id} retry count exceeded maximum')
+        else:
+          # This is very common if multiple instances are sending final update
+          # simultaneously.
+          native.LogInfo(f'{err}, breakpoint: {bp_id}')
+      except socket.error as err:
+        if retry_count < self.max_transmit_attempts - 1:
+          native.LogInfo(f'Socket error {err.errno} while sending breakpoint '
+                         f'{bp_id} update: {traceback.format_exc()}')
+          retry_list.append((breakpoint_data, retry_count + 1))
+        else:
+          native.LogWarning(f'Breakpoint {bp_id} retry count exceeded maximum')
+          # Socket errors shouldn't persist like this; reconnect.
+          #reconnect = True
+      except BaseException:
+        native.LogWarning(f'Fatal error sending breakpoint {bp_id} update: '
+                          f'{traceback.format_exc()}')
+
+    self._transmission_queue.extend(retry_list)
+
+    if not self._transmission_queue:
+      self.update_backoff.Succeeded()
+      # Nothing to send, wait until next breakpoint update.
+      return None
+    else:
+      return self.update_backoff.Failed()
+
+  def _GetDebuggee(self):
+    """Builds the debuggee structure."""
+    major_version = 'v' + version.__version__.split('.', maxsplit=1)[0]
+    python_version = ''.join(platform.python_version().split('.')[:2])
+    agent_version = f'google.com/python{python_version}-gcp/{major_version}'
+
+    debuggee = {
+        'description': self._GetDebuggeeDescription(),
+        'labels': self._debuggee_labels,
+        'agentVersion': agent_version,
+        'canaryMode': self._canary_mode,
```

Comment:
Is there a reason to keep canary in the firebase version?

Reply:
Absolutely not.  Thanks, I'll remove that.